Retain the existing site list data when you start searching

p-jackson opened the pull request:
<!--
Link a related Github/Linear issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword. Consider using "part of" instead.
-->



## Proposed Changes

The site list search box doesn't just search visible sites, it also searches sites which have been hidden. That is the site list empties when you start searching: it changes from a `?site_visibility=visible` fetch, to a `?site_visibility=all` fetch.

This PR changes the site list component so that retains the old data in the dataview when the new fetch is initiated. It is much less disruptive for most users who probably aren't looking for a hidden site.

**Before**


https://github.com/user-attachments/assets/a7564c2d-31d6-4bb4-b9fa-4d12090902eb



**After**


https://github.com/user-attachments/assets/048b3c55-a352-49b3-a0b9-2e8cdfa72f5c


**When you _do_ have a hidden site**


https://github.com/user-attachments/assets/b0300f97-f559-42ed-ab9c-805acdaadb3c




## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

Related to some product feedback pc4f5j-5KC-p2#comment-4989

While we can't improve the list performance until we've switched to elastic search, we can still improve the perceived performance.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Search using the site list in v2 and the backport
* You may want to clear out the react query cache between tests so you are confident of when it is fetching new data
* You can manage your hidden sites here: https://dashboard.wordpress.com/wp-admin/index.php?page=my-blogs

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you tested accessibility for your changes? Ensure the feature remains usable with various user agents (e.g., browsers), interfaces (e.g., keyboard navigation), and assistive technologies (e.g., screen readers) (PCYsg-S3g-p2).
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

p-jackson committed:
Retain the existing site list data when you start searching

arthur791004 reviewed (approved):
Nice 👍

arthur791004 commented:
I guess we can simply pass `isFetching` to the DataViews to resolve this issue

fushar reviewed (commented):
What if we also preload the "`?site_visibility=all` fetch" here 

https://github.com/Automattic/wp-calypso/blob/0cfdd7ea816e457ccccb11fa952da12572adce5d/client/dashboard/app/router/sites.tsx#L43-L45

p-jackson committed:
Show spinner when the placeholder data is empty

p-jackson commented:
> I guess we can simply pass isFetching or isPlaceholderData to the isLoading prop of DataViews to resolve this issue

Good idea. I've updated the PR and updated the video in the description to show what it does now when the search doesn't match any of the visible sites. This feels much better.

I still think the idea of a "fetching is happening in the background" indicator would be good too, because now if your search matches some of your visible sites, but you are actually looking for your hidden site, you have no indication that you need to just wait a couple of seconds before it will pop up.

> What if we also preload the "?site_visibility=all fetch" here

It seems kinda heavy to do two requests IMO. A lot of the time I just land on the site list because it is the default view, I'm not looking to load gobs of data when I just wanna navigate somewhere else.

If we _did_ want to have all that data client side then it would be better to just query `site_visibility=all` from the beginning and filter out the hidden sites client side.
